knimin/lib/vioscreen.py

```diff
@@ -0,0 +1,446 @@
+            The data that is getting associated with the survey_id
+
+        Return
+        ------
+        list of dict
+            The newly formatted data containing the username in each row
+        """
+        dat = []
+        for entry in payload:
+            entry['survey_id'] = username
+            dat.append(entry)
+        return dat
+
+    def get_session_data(self, session_id, endpoint):
+        """Pulls data from the vioscreen API based on 
+           a specific session ID and session type(ex. 'foodcomponents')
+
+        Parameters
+        ----------
+        session_id: str
+            Session ID that data is being requested for
+        endpoint: str
+            Name of the session type that data is being requested for
+
+        Return
+        ------
+        dict:
+            Food frequency questionnaire data
+        """
+        if not self._headers:
+            self._headers = {'Accept': 'application/json',
+                             'Authorization': 'Bearer %s' % self.get_token()}
+        return self.get('https://api.viocare.com/%s/sessions/%s/%s' %
+                         (self._key, session_id, endpoint),
+                         headers=self._headers)
+
+    def sync_vioscreen(self, user_ids=None):
+        """Pulls data from the vioscreen API and stores
+           the data into the AG database
+
+        Parameters
+        ----------
+        user_ids: set of str
+            Set of user_ids (identical to survey_ids) that
+            are needed to have their data pulled. Default None (syncs all)
+        """
+        if not self._headers:
+            self._headers = {'Accept': 'application/json',
+                             'Authorization': 'Bearer %s' % self.get_token()}
+        if not self._users:
+            self._users = self.get_users()
+
+        if user_ids:
+            if type(user_ids) is not set:
+                raise TypeError('user_ids should be type set')
+        else:
+            user_ids = {x['username'] for x in self._users['users']}
+
+        # takes all survey IDs from vio_screen survey info and filters
+        # only ones that do not have their data in the ag database
+        ids_to_sync = self.get_vio_survey_ids_not_in_ag(user_ids)
+
+        # gets all survey info of ids_to_sync and updates users
+        users = {'users': []}
+        for user in self._users['users']:
+            if user['username'] in ids_to_sync:
+                users['users'].append(user)
+
+        # gets list of surveys in AG database along with their statuses
+        survey_ids = self.get_init_surveys()
+
+        for user in users['users']:
+            username = user['username']
+
+            try:
+                session_data = self.get('https://api.viocare.com/%s/users/%s/sessions'
+                                        % (self._key, username), headers=self._headers)
+            except ValueError:
+                # I don't understand this, but "JDebelius" does not exist.
```

Comment:
Weird

I wonder if that username should be filtered out by `get_users`?

Reply:
I've never seen this error occur in my many tests.